scipy/optimize/_qap.py

```diff
@@ -0,0 +1,722 @@
+        Setting `maximize` to ``True`` solves the Graph Matching Problem (GMP)
+        rather than the Quadratic Assingnment Problem (QAP).
+
+    rng : {None, int, `~np.random.RandomState`, `~np.random.Generator`}
+        This parameter defines the object to use for drawing random
+        variates.
+        If `rng` is ``None`` the `~np.random.RandomState` singleton is
+        used.
+        If `rng` is an int, a new ``RandomState`` instance is used,
+        seeded with `rng`.
+        If `rng` is already a ``RandomState`` or ``Generator``
+        instance, then that object is used.
+        Default is None.
+
+    Returns
+    -------
+    res : OptimizeResult
+        A :class:`scipy.optimize.OptimizeResult` containing the following
+        fields.
+
+        col_ind : 1-D array
+            An array of column indices corresponding with the best
+            permutation of the nodes of `B` found.
+        score : float
+            The corresponding value of the objective function.
+        nit : int
+            The number of iterations performed during optimization.
+
+    Notes
+    -----
+    This is a greedy algorithm that works similarly to bubble sort: beginning
+    with an initial permutation, it iteratively swaps pairs of indices to
+    improve the objective function until no such improvements are possible.
+
+    References
+    ----------
+    .. [3] "2-opt," Wikipedia.
+           https://en.wikipedia.org/wiki/2-opt
+    """
+
+    def calc_score(A, B, perm):
+        # equivalent for optimization per FAQ paper [1]; avoids matmul
+        k = np.argsort(perm)
+        return -((A[:, k] - B[perm, :])**2).sum()
+
+    rng = check_random_state(rng)
+    A, B, partial_match = _common_input_validation(
+            A, B, partial_match, maximize)
+
+    N = len(A)
+
+    if partial_match.size:
+        r, c = partial_match.T
+        fixed_rows = np.zeros(N, dtype=bool)
+        fixed_cols = np.zeros(N, dtype=bool)
+        fixed_rows[r] = True
+        fixed_cols[c] = True
+
+        perm = np.zeros(N, dtype=int)
+        perm[fixed_rows] = c
+        perm[~fixed_rows] = rng.permutation(np.arange(N)[~fixed_cols])
+    else:
+        perm = rng.permutation(np.arange(N))
+
+    best_score = calc_score(A, B, perm)
+
+    better = operator.gt if maximize else operator.lt
+    n_iter = 0
+    done = False
+    while not done:
+        # equivalent to nested for loops i in range(N), j in range(i, N)
+        for i, j in itertools.combinations_with_replacement(range(N), 2):
+            n_iter += 1
+            perm[i], perm[j] = perm[j], perm[i]
+            score = calc_score(A, B, perm)
+            if better(score, best_score):
+                best_score = score
+                break
+            # faster to swap back than to create a new list every time
+            perm[i], perm[j] = perm[j], perm[i]
+        else:  # no swaps made
+            done = True
+
+    best_score = np.trace(A @ B[np.ix_(perm, perm)].T)
```

Comment:
Does this need to be recalculated? If so, can we use `calc_score`?

Reply:
Yes it does, and no you can't. `calc_score` is not the same number, but it's equivalent for the sake of optimization.

Reply:
OK. I'm on the same page.

Reply:
I have added a suggestion to `calc_score` which should mean that we don't need to recalculate here.

Reply:
Done. Good idea. I was trying to escape matmul but that is clearly the better way.